rbm_pytorch.py

```diff
@@ -302,7 +302,15 @@ def free_energy(self, v, beta=1.0):
         return (hidden_term + vbias_term)*beta  # mean along the batches
 
     def free_energy_batch_mean(self, v, beta = 1.0):
-        return self.free_energy.mean()
+        return self.free_energy(v).mean()
+
+
+    def log_likelihood(self, v, beta=1.0):
+        # compute the log-likelihood
+        logz , logz_up, logz_down = self.annealed_importance_sampling(1, 10000, 100)
+        return self.free_energy_batch_mean(v) - logz
+
```

Comment:
So I think here we should instead have 

```
return sum(self.free_energy(v) - logz)
```

which gives log likelihood for a set of visible layers v, I think this is correct according to my email and eq(4) of http://image.diku.dk/igel/paper/AItRBM-proof.pdf and also the added bonus is unlike taking the mean this actually gives the same answer when done separately across batches and then summed at end of epoch or just done at the end of epoch

Reply:
actually I just realised that this is exactly the same thing up to a factor of N, where N is the number of visible states given to `log_likelihood` In other words this wouldn't change the result if done at the end of the epoch across all states, but would change how you would have to sum up individual contributions if calculated from batches

Reply:
so we want to implement eq.(4), say for each batch, right?
Then i would try

'''
return  (self.free_energy(v).sum() - logz*N)
'''
where N is the batch size, and logz is computed only once at the beginning or at the end of the batch, not at every update of the parameters, to speed up the process.

However this is not

'''
return  (self.free_energy(v) - logz)*N
'''
right?
In this way we would be assuming the free energy to be the same for every state in the batch (as we are doing with logz). Does it make sense?

Reply:
Ok the top expression I agree with however if you take a scalar `logZ` away from a vector `free_energy` then it takes it away from all components and so

```
self.free_energy(v).sum() - logz*N
```

is exactly the same as

```
sum(self.free_energy(v) - logz)
```

the second expression doesn't quite work because you have the vector of free energies for each visible state and then take away logz from each one then times by N. This will return a vector of (free_energy - logz)*N

Reply:
if you make me a collaborator (in the settings tab for this repo) then I can push changes which might be good.

The thing I was saying with second comment is that 

free_energy.mean = free_energy.sum/N

so free energy.mean - logz = (free_energy.sum() - N*logz)/N

so is basically the same as eq.4 but all divided by N

Reply:
>where N is the batch size, and logz is computed only once at the beginning or at the end of >the batch, not at every update of the parameters, to speed up the process.

I don't understand this part, at the moment you compute logz once per epoch, the weights get updated once per batch. I think calculating logz every batch might take too long, probably just worth doing it once per epoch.. In that sense what you are currently doing in the code looks fine to me because now

```
free_energy.mean() - logz
```

is the same as 

```
sum(free_energy - logz)/N_epoch
```

in other words each epoch we find LL according to eq (4) and then divide by N_epoch which is just a constant (100000 in our case) plotting how this LL evolves with epoch should tell us exactly what we want because it should tend to a maximum!

Reply:
Yes sorry, i was confused between batches and epochs.
I agree, that' s what is currently implemented in the code, right?
By the way you should be a collaborator now